### ISSUE TITLE
let_and_return false positive in presence of `[cfg]`ed out blocks

### ISSUE THREAD

**Issue by RealKC** [rust-lang/rust-clippy#9150]
### Summary

clippy emits a `let_and_return` warning when between the let and return there is only a `cfg`ed out block

### Lint Name

let_and_return

### Reproducer

I tried this code:

```rust
fn _uwu() -> bool {
    let x = false;
    
    #[cfg(feature="owo")]
    {
        println!("Some more complex code here");
    }
    
    x
}
```

I saw this happen:

```
warning: returning the result of a `let` binding from a block
 --> src/lib.rs:9:5
  |
2 |     let x = false;
  |     -------------- unnecessary `let` binding
...
9 |     x
  |     ^
  |
  = note: `#[warn(clippy::let_and_return)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_and_return
help: return the expression directly
  |
2 ~     
3 |     
...
8 |     
9 ~     false
  |

```

I expected to see this happen:
clippy shouldn't emit a warning here

### Version

```text
rustc 1.64.0-nightly (06754d885 2022-07-08)
binary: rustc
commit-hash: 06754d8852bea286a3a76d373ccd17e66afb5a8b
commit-date: 2022-07-08
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.6
```

**Comment by alex-semenyuk** [comment]
It has the same behaviour if there is no `cfg` block. Why do you think it's false positive?

**Comment by RealKC** [comment]
> It has the same behaviour if there is no `cfg` block. Why do you think it's false positive?

I think it's a false positive because only removing the `#[cfg(feature="owo")]` line stops the warning from being emitted.

**Comment by xFrednet** [comment]
I believe that the idea of the lint suggestion is, that you move the variable value to the return statement like this:

```rs
fn _uwu() -> bool {
    #[cfg(feature="owo")]
    {
        println!("Some more complex code here");
    }
    
    false
}
```

or to notify the user that it hasn't been modified. Maybe we could add a note to the lint message to explain this better :upside_down_face:

**Comment by RealKC** [comment]
I do understand the idea of the lint and generally agree with it, but IMO either clippy should emit a lint for this code too

```rust
fn _uwu() -> bool {
    let x = false;
    
    {
        println!("Some more complex code here");
    }
    
    x
}
```

which it currently doesn't, or not emit a lint for the code in the OP.

**Comment by wr7** [comment]
To add to this, this can change the behavior code when `cargo clippy --fix` is run. 
For example, in the following code, `foobar` normally prints `foo! bar! ` when run with the `bar` feature:
```rust
#[cfg(feature = "bar")]
fn bar() {
    print!("bar! ");
}

fn foo() -> i32 {
    print!("foo! ");
    4
}

fn foobar() -> i32 {
    let foo = foo();

    #[cfg(feature = "bar")]
    bar();

    foo
}
```
However, after running `cargo clippy --fix`, it prints `bar! foo! ` because the function `foobar` is changed to:
```rust
fn foobar() -> i32 {
    #[cfg(feature = "bar")]
    bar();

    foo()
}
```

**Comment by andersonjwan** [comment]
I am facing the same issue as well. It would be nice for `clippy` to contextualize to the `#[cfg(feature = "...")]` macros, so as to not issue incorrect logical warnings as these situations are intentional.

**Comment by stepantubanov** [comment]
```rust
pub fn foo(x: &mut i32) {
    *x += 1;
}

pub fn bar() -> i32 {
    let x = 1;

    #[cfg(feature = "bar")]
    foo(&mut x);

    x
}
```
clippy suggestion is to return `1` from `bar()`

**Comment by y21** [comment]
Clippy doesn't see the cfg attributes since they're expanded to nothing when the lint runs, but maybe we can retokenize everything from at the end of the local to the start of the tail expression and check if the token sequence `#[` is there, and then avoid emitting a warning? 🤔
